contracts/crosschain/wormhole/WormholeGatewayAdapter.sol

```diff
@@ -0,0 +1,325 @@
+    function registerRemoteGateway(bytes calldata remote) public virtual /*onlyOwner in registerRemoteGateway*/ {
+        (uint256 chainId, address addr) = remote.parseEvmV1Calldata();
+        registerRemoteGateway(chainId, addr);
+    }
+
+    /// @dev Registers the address of a remote gateway (EVM version).
+    function registerRemoteGateway(uint256 chainId, address addr) public virtual onlyOwner {
+        require(supportedChain(chainId), UnsupportedChainId(chainId));
+        require(_remoteGateways[chainId] == address(0), RemoteGatewayAlreadyRegistered(chainId));
+        _remoteGateways[chainId] = addr;
+        emit RegisteredRemoteGateway(chainId, addr);
+    }
+
+    /// @inheritdoc IERC7786GatewaySource
+    function supportsAttribute(bytes4 selector) public pure returns (bool) {
+        return selector == IERC7786Attributes.requestRelay.selector;
+    }
+
+    /// @inheritdoc IERC7786GatewaySource
+    function sendMessage(
+        bytes calldata recipient, // Binary Interoperable Address
+        bytes calldata payload,
+        bytes[] calldata attributes
+    ) external payable returns (bytes32 sendId) {
+        for (uint256 i = 0; i < attributes.length; ++i) {
+            bytes4 selector = attributes[i].length < 0x04 ? bytes4(0) : bytes4(attributes[i]);
+            require(supportsAttribute(selector), UnsupportedAttribute(selector));
+        }
+
+        // We need a unique message identifier.
+        bytes32 msgId = bytes32(++_lastMsgId);
+
+        // Case 1. We don't have a requestRelay attribute. The message is saved, waiting for a call to requestRelay.
+        if (attributes.length == 0) {
+            // Note: this reverts with UnsupportedChainId if the recipient is not on a supported chain.
+            // No real need to check the return value.
+            getRemoteGateway(recipient);
+
+            // Store the message for future execution in {requestRelay}
+            _pending[msgId] = PendingMessage(true, msg.sender, msg.value, recipient, payload);
+
+            emit MessageSent(
+                msgId,
+                InteroperableAddress.formatEvmV1(block.chainid, msg.sender),
+                recipient,
+                payload,
+                msg.value,
+                attributes
+            );
+
+            return msgId;
+        }
+
+        // Case 2. We have a requestRelay attribute.
+        if (attributes.length == 1) {
+            // Parse the attribute details
+            (bool success, uint256 receiverValue, uint256 gasLimit, address refundRecipient) = ERC7786Attributes
+                .tryDecodeRequestRelay(attributes[0]);
+            require(success, InvalidAttributeEncoding(attributes[0]));
+
+            // Send the message
+            _sendMessage(msgId, recipient, payload, msg.sender, msg.value, receiverValue, gasLimit, refundRecipient);
+
+            emit MessageSent(
+                0,
```

Comment:
If we have an id we might as well return it? It can be used to track the MessageRelayed event later.

Reply:
Yes, returning an id means post processing _may_ be necessary.

> I would encourage returning zero is we have no usage for the value.

I was missing that if we return zero there won't be a MessageRelayed event. So it's true there is no usage for it in this case. Fine to leave it as is.